### PR TITLE
Fix 507 complete mapping expansion

### PR DESCRIPTION
**Pull request by brynrhodes**
Note that this PR does not completely address the mapping expansion, and it has no impact on models that do not use the mapping attributes (such as target). 

I'm submitting it for merge to the main branch so I can subsequently pull everything in 1.4 through to the 1.5 develop branch and will continue working on this issue there. I just need to release 1.4 and this feature is not going to make it in 1.4.

### CODE REVIEW

File: Src/grammar/targetmap.g4 [hunk 1]
```diff
@@ -0,0 +1,108 @@
+grammar targetmap;
+
+/*
+Target mapping language to support mapping expressions for primitives, choices, extensions, and slicing:
+
+// Primitives
+%value.value
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/grammar/targetmap.g4:7]: It's not really `$this`, it's more a parameter/place-holder, and the challenge with using $this is that some of the slice maps are actually FHIRPath and do already contain $this, so I can't use that as a pre-processing directive.

File: Src/java/quick/src/main/resources/org/hl7/fhir/fhir-modelinfo-4.0.1.xml [hunk 2]
```diff
@@ -2957,129 +2957,6 @@
...
-        <ns4:element name="constraint">
-            <ns4:elementTypeSpecifier xsi:type="ns4:ListTypeSpecifier" elementType="FHIR.ElementDefinition.Constraint"/>
-        </ns4:element>
-        <ns4:element name="mustSupport" elementType="FHIR.boolean"/>
-        <ns4:element name="isModifierReason" elementType="FHIR.string"/>
-        <ns4:element name="binding" elementType="FHIR.ElementDefinition.Binding"/>
-        <ns4:element name="mapping">
-            <ns4:elementTypeSpecifier xsi:type="ns4:ListTypeSpecifier" elementType="FHIR.ElementDefinition.Mapping"/>
-        </ns4:element>
-    </ns4:typeInfo>
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/java/quick/src/main/resources/org/hl7/fhir/fhir-modelinfo-4.0.1.xml:3082]: The constructs being removed here are due to tightening up the generation. The "Data Element profile on Element Definition" class, for example, is a profile class that shouldn't have been in the 4.0.1 model info. We do have CQL using 4.0.1, and that is the preferred model moving forward, but we preserve the 4.0.0 model to ensure backwards compatibility.

File: Src/grammar/targetmap.g4 [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+QICoreHelpers.ToConcept(%value)
+
+// Choices
+System.DateTime:%value.value;;QICore.Timing:null
+QICore.Reference:null;System.Concept:QICoreHelpers.ToConcept(%value)
+
+// Extensions
+QICoreHelpers.ToConcept(%value.extension[url=http://hl7.org/fhir/us/qicore/StructureDefinition/qicore-doNotPerformReason].value)
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/grammar/targetmap.g4:15]: Noted in the TODO

File: Src/grammar/targetmap.g4 [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+%value.modifierExtension[url=http://hl7.org/fhir/StructureDefinition/request-doNotPerform].value.value
+
+
+// Slicing
+QICoreHelpers.ToConcept(%value.category[coding.system=http://terminology.hl7.org/CodeSystem/observation-category,coding.code=vital-signs].value)
+QICoreHelpers.ToCode(%value.code.coding[system=http://loinc.org,code=39156-5].value)
+%value.component[code.coding.system=http://loinc.org,code.coding.code=8480-6]
+%value.component[code.coding.system=http://loinc.org,code.coding.code=8462-4]
+*/
+
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/grammar/targetmap.g4:25]: Noted in the TODO

File: Src/grammar/targetmap.g4 [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+System.DateTime:%value.value;;QICore.Timing:null
+QICore.Reference:null;System.Concept:QICoreHelpers.ToConcept(%value)
+
+// Extensions
+QICoreHelpers.ToConcept(%value.extension[url=http://hl7.org/fhir/us/qicore/StructureDefinition/qicore-doNotPerformReason].value)
+%value.modifierExtension[url=http://hl7.org/fhir/StructureDefinition/request-doNotPerform].value.value
+
+
+// Slicing
+QICoreHelpers.ToConcept(%value.category[coding.system=http://terminology.hl7.org/CodeSystem/observation-category,coding.code=vital-signs].value)
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/grammar/targetmap.g4:20]: Noted in the TODO

File: Src/grammar/targetmap.g4 [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+choiceMapItem
+  : (qualifiedIdentifier ':' targetMap)?
+  ;
+
+targetIndex
+  : targetIndexItem (',' targetIndexItem)*
+  ;
+
+targetIndexItem
+  : identifier '=' STRING
```
[the remark this answers is not in the document]
Comment (brynrhodes) [reply Src/grammar/targetmap.g4:51]: Noted in the TODO

**Comment by brynrhodes** [comment]
Thank you as always for the review Chris! I added your consideration questions to the targetmap.g4 grammar so we can continue those discussions in the next release. As for why I think this needs to be in the 1.4.9 release, what's here is functional for a lot of mapping functionality (primitives and extensions, and some slicing. See the [regression tests branch](https://github.com/DBCG/cql-evaluator/blob/fix-13-regression-tests/cli/src/test/resources) of the CQL Evaluator for tests that currently pass. But there are some slicing patterns that don't work (see the only uncommented test in [the uscore test](https://github.com/DBCG/cql-evaluator/blob/fix-13-regression-tests/cli/src/test/resources/uscore/TestUSCore.cql#L83). But I don't want to go backwards on progress that has already been made, and not bringing this in to the 1.4.9 release would do that.

**Comment by cmoesel** [comment]
OK.  Makes sense.  Thanks for the explanation!

**Comment by cmoesel** [comment]
OK.  Makes sense.  Thanks for the explanation!

**Pull request merged**